src/drivers/imx/esai.c

```diff
@@ -3,38 +3,450 @@
+// Author: Paul Olaru <paul.olaru@nxp.com>
 
+#include <sof/audio/component.h>
+#include <sof/drivers/edma.h>
 #include <sof/drivers/esai.h>
+#include <sof/lib/alloc.h>
 #include <sof/lib/dai.h>
 #include <sof/lib/dma.h>
 #include <ipc/dai.h>
+#include <ipc/topology.h>
+#include <errno.h>
+#include <stdint.h>
+
+struct esai_pdata {
+	struct {
+		uint32_t ecr;
+		uint32_t tfcr;
+		uint32_t rfcr;
+		uint32_t saicr;
+		uint32_t tcr;
+		uint32_t tccr;
+		uint32_t rcr;
+		uint32_t rccr;
+		uint32_t tsma;
+		uint32_t tsmb;
+		uint32_t rsma;
+		uint32_t rsmb;
+		uint32_t prrc;
+		uint32_t pcrc;
+	} regs;
+	/* TODO set the config here, prepare the registers and defer the
+	 * actual updating of the HW registers to the start() and stop()
+	 * triggers
+	 */
+};
 
 static int esai_context_store(struct dai *dai)
 {
+	struct esai_pdata *pdata = dai_get_drvdata(dai);
+
+	if (!pdata)
+		return -EINVAL;
+
+	pdata->regs.ecr = dai_read(dai, REG_ESAI_ECR);
+	pdata->regs.tfcr = dai_read(dai, REG_ESAI_TFCR);
+	pdata->regs.rfcr = dai_read(dai, REG_ESAI_RFCR);
+	pdata->regs.saicr = dai_read(dai, REG_ESAI_SAICR);
+	pdata->regs.tcr = dai_read(dai, REG_ESAI_TCR);
+	pdata->regs.tccr = dai_read(dai, REG_ESAI_TCCR);
+	pdata->regs.rcr = dai_read(dai, REG_ESAI_RCR);
+	pdata->regs.rccr = dai_read(dai, REG_ESAI_RCCR);
+	pdata->regs.tsma = dai_read(dai, REG_ESAI_TSMA);
+	pdata->regs.tsmb = dai_read(dai, REG_ESAI_TSMB);
+	pdata->regs.rsma = dai_read(dai, REG_ESAI_RSMA);
+	pdata->regs.rsmb = dai_read(dai, REG_ESAI_RSMB);
+	pdata->regs.prrc = dai_read(dai, REG_ESAI_PRRC);
+	pdata->regs.pcrc = dai_read(dai, REG_ESAI_PCRC);
+
 	return 0;
 }
 
 static int esai_context_restore(struct dai *dai)
 {
+	struct esai_pdata *pdata = dai_get_drvdata(dai);
+
+	if (!pdata)
+		return -EINVAL;
+
+	dai_write(dai, REG_ESAI_ECR, ESAI_ECR_ERST);
+	dai_write(dai, REG_ESAI_ECR, ESAI_ECR_ESAIEN);
+	dai_write(dai, REG_ESAI_TFCR, pdata->regs.tfcr);
+	dai_write(dai, REG_ESAI_RFCR, pdata->regs.rfcr);
+	dai_write(dai, REG_ESAI_SAICR, pdata->regs.saicr);
+	dai_write(dai, REG_ESAI_TCCR, pdata->regs.tccr);
+	dai_write(dai, REG_ESAI_RCCR, pdata->regs.rccr);
+	dai_write(dai, REG_ESAI_TSMA, pdata->regs.tsma);
+	dai_write(dai, REG_ESAI_TSMB, pdata->regs.tsmb);
+	dai_write(dai, REG_ESAI_RSMA, pdata->regs.rsma);
+	dai_write(dai, REG_ESAI_RSMB, pdata->regs.rsmb);
+	dai_write(dai, REG_ESAI_PRRC, pdata->regs.prrc);
+	dai_write(dai, REG_ESAI_PCRC, pdata->regs.pcrc);
+	dai_write(dai, REG_ESAI_TCR, pdata->regs.tcr);
+	dai_write(dai, REG_ESAI_RCR, pdata->regs.rcr);
+	dai_write(dai, REG_ESAI_ECR, pdata->regs.ecr);
+
 	return 0;
 }
 
 static inline int esai_set_config(struct dai *dai,
 				 struct sof_ipc_dai_config *config)
 {
+	uint32_t xcr = 0, xccr = 0, mask;
+
+	tracev_esai("ESAI: set_config format 0x%4x (explaining)",
+		    config->format);
+	switch (config->format & SOF_DAI_FMT_FORMAT_MASK) {
+	case SOF_DAI_FMT_I2S:
+		tracev_esai("ESAI: format I2S");
+		/* Data on rising edge of bclk, frame low, 1clk before
```

Comment:
Remove this specific traces as it makes code harder to read. Also please remove (explaining) from toplevel trace.